### PR TITLE
Sync Hierarchy optimisation

### PR DESCRIPTION
**Pull request by e-strokov**
Frozen property added to optimise SyncHierarchy.
Sync skips all frozen entities and their sub-hierarchy.
Any changes to transform unfreeze this entity with children and entire parent chain to a root.
The following sync pass switches back to frozen.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/scene/graph-node.js [hunk 1]
```diff
@@ -911,12 +913,27 @@ Object.assign(pc, function () {
             }
         },
 
+        _unfreezeParentToRoot: function () {
+            var p = this._parent;
+            while (p) {
```
Comment (daredevildave) [inline src/scene/graph-node.js]: Can we exit the loop here if `p._frozen === false`? That means we already have a path up to the root that is unfrozen?

File: src/scene/graph-node.js [hunk 2]
```diff
@@ -35,6 +35,8 @@ Object.assign(pc, function () {
         this._dirtyLocal = false;
         this._aabbVer = 0;
 
+        this._frozen = false;
```
Comment (daredevildave) [inline src/scene/graph-node.js]: Would be good to have a definition of what frozen is doing. Expecially in contrast to dirtyWorld/dirtyLocal.

Comment (daredevildave) [reply src/scene/graph-node.js]: By definition, I mean a comment.

**Comment by lvenerosy** [comment]
Coming from https://github.com/playcanvas/engine/issues/1517, we definitely get a performance boost with this branch! But we have an image element without texture/sprite/material set, just the color, which does not get resized properly. I will investigate to see if it comes from our code.

**Comment by willeastcott** [comment]
Interesting. I think the untextured image element may be unrelated to this change. We've had a similar report from another developer recently. cc @daredevildave

**Comment by lvenerosy** [comment]
There were actually 2 different issues that I thought might have the same origin. One of them was totally our fault because someone, definitely not me^^, thought it was a good idea to tween straight over `getLocalScale` even though the comment says it should be considered read only. So fixing that made the first issue go away on both master and this branch.

The other issue though still persists on this branch but not master. We use image elements for icons overlays so all of them should have the same behavior but a single one is blown out and takes up the whole bottom right quarter of the screen, and scaling it does not change its size. It happens on a single one of our icons so we are most likely at fault again.

**Comment by lvenerosy** [comment]
It was our mistake indeed. This branch works fine with our project!

File: src/scene/graph-node.js (hunk 1 above)
Comment (e-strokov) [reply src/scene/graph-node.js]: That was my initial idea, but unfortunately there is a case when after some cloning-reparenting operations sync becomes broken. I had to implement '_dirtifyWorldInternal' method to minimise performance impact.

File: src/scene/graph-node.js (hunk 2 above)
Comment (e-strokov) [reply src/scene/graph-node.js]: Yes, I will add the description.

**Comment by gameatp** [comment]
Is there a better way to use dirty in this way to know when I should do it manually?
Or I just add an object and execute the _dirtifyWorldInternational method once




------------------ 原始邮件 ------------------
发件人: "e-strokov"<notifications@github.com>;
发送时间: 2019年6月3日(星期一) 晚上6:25
收件人: "playcanvas/engine"<engine@noreply.github.com>;
抄送: "Subscribed"<subscribed@noreply.github.com>;
主题: Re: [playcanvas/engine] Sync Hierarchy optimisation (#1576)




@e-strokov commented on this pull request.
 
 
In src/scene/graph-node.js:
 > @@ -911,12 +913,27 @@ Object.assign(pc, function () {              }          },   +        _unfreezeParentToRoot: function () { +            var p = this._parent; +            while (p) {  
That was my initial idea, but unfortunately there is a case when after some cloning-reparenting operations sync becomes broken. I had to implement '_dirtifyWorldInternal' method to minimise performance impact.

**Comment by malDuffin** [comment]
It would be useful to have the equivalent of this built into this fix also, so that custom code could check if something else was needed on update ( eg for updating custom bounding volumes )

https://docs.unity3d.com/ScriptReference/Transform-hasChanged.html

**Pull request merged**